pmlc/conversion/gpu/gpu_to_vulkan.cc

```diff
@@ -237,10 +262,86 @@ Value ConvertGpuLaunchFuncToVulkanCalls::createEntryPointNameConstant(
                                   shaderName, LLVM::Linkage::Internal);
 }
 
+llvm::SmallSet<Operation *, 4>
+ConvertGpuLaunchFuncToVulkanCalls::getExternalDependentOperations(
+    Value &value) {
+  llvm::SmallSet<Operation *, 4> operations;
+  llvm::SetVector<OpOperand *> uses;
+
+  operations.insert(value.getDefiningOp());
+  for (auto &use : value.getUses()) {
+    uses.insert(&use);
+  }
+  while (!uses.empty()) {
+    auto owner = uses.back()->getOwner();
+    operations.insert(owner);
+    uses.pop_back();
+    for (auto opResult : owner->getOpResults()) {
+      for (auto &use : opResult.getUses()) {
+        uses.insert(&use);
+      }
+    }
+  }
+
+  for (auto operation : operations) {
+    if (isInternalOperation(operation)) {
+      operations.erase(operation);
+    }
+  }
+  return operations;
+}
+
+bool ConvertGpuLaunchFuncToVulkanCalls::isInternalOperation(Operation *op) {
+  // Buffer related calls are all mlir::CallOp, does not need to check
+  // LLVM::CallOp
+  auto callOp = llvm::dyn_cast<mlir::CallOp>(op);
+  if (callOp && (callOp.callee() == kBindAllBuffers ||
+                 optionalSymbols.count(callOp.callee()))) {
+    return true;
+  }
+
+  auto allocOp = llvm::dyn_cast<AllocOp>(op);
+  auto memRefCastOp = llvm::dyn_cast<MemRefCastOp>(op);
+  if (allocOp || memRefCastOp) {
+    return true;
+  }
+  return false;
+}
+
+uint32_t
+ConvertGpuLaunchFuncToVulkanCalls::getBufferCopyMode(mlir::CallOp &callOp,
+                                                     Value &buffer) {
+  uint32_t copyMode = kBufferCopyModeInit;
+  if (buffer.isa<BlockArgument>()) {
+    copyMode |= (kBufferCopyModeHostToDevice | kBufferCopyModeDeviceToHost);
+    return copyMode;
+  }
+
+  auto operationDeps = getExternalDependentOperations(buffer);
+  auto currentBlock = callOp.getOperation()->getBlock();
+  for (auto op : operationDeps) {
+    if (op->getBlock() == currentBlock) {
+      if (op->isBeforeInBlock(callOp.getOperation())) {
+        copyMode |= kBufferCopyModeHostToDevice;
+      } else if (deinitVulkan.getOperation()->isBeforeInBlock(op)) {
+        copyMode |= kBufferCopyModeDeviceToHost;
+      } else {
+        callOp.emitWarning("A host side buffer is used after copied to "
+                           "device and before device returns.");
+      }
+    } else {
+      copyMode |= (kBufferCopyModeHostToDevice | kBufferCopyModeDeviceToHost);
```

Comment:
You can `return copyMode` here as well.  Once we know there is a dependency outside the current block we are done.

Reply:
I think right now for debugging purposes, we have to check all dependencies to see if some of them changed host side buffers after copied to device. 
If we fixed this issue or we do not want to check here, we can early return copyMode.